### PR TITLE
Typing utils

### PR DESCRIPTION
**Pull request by PietroPasotti**
Typing for the remaining "small" modules of ops:

log.py
version.py
jujuversion.py
main.py

### CODE REVIEW

File: ops/jujuversion.py [hunk 1]
```diff
@@ -55,28 +56,29 @@ def __repr__(self):
             s += '.{}'.format(self.build)
         return s
 
-    def __eq__(self, other):
+    def __eq__(self, other: object) -> bool:
```
Comment (jnsgruk) [inline ops/jujuversion.py:59]: Would the following suggested change not obviate the need for the `cast`? (applies to the `__lt__` dunder too)

```suggestion
    def __eq__(self, other: Union[str, "JujuVersion"]) -> bool:
```

**Review by jnsgruk — APPROVED** [review]
One comment, but otherwise looks good.

File: pyproject.toml [hunk 2]
```diff
@@ -34,9 +34,8 @@ docstring-convention = "google"
 [tool.pyright]
 exclude = ["ops/_vendor", ".git", "__pycache__", ".tox", "build", "test",
     "ops/lib", "ops/_private", "dist", "*.egg_info", "venv", "ops/charm.py",
-    "ops/framework.py", "ops/jujuversion.py", "ops/log.py", "ops/main.py",
-    "ops/model.py", "ops/pebble.py", "ops/testing.py",
-    "ops/version.py", "setup.py", "ops/__init__.py"]
-pythonVersion = 3.5 # check no python > 3.5 features are used
+    "ops/framework.py", "ops/main.py", "ops/model.py", "ops/pebble.py",
+    "ops/testing.py", "ops/__init__.py"]
+pythonVersion = "3.5" # check no python > 3.5 features are used
```
Comment (sed-i) [inline pyproject.toml:39]: The analogous for mypy, `mypy --python-version 3.5`, is not sufficient (gives false negatives): we had to [run with py3.5](https://github.com/canonical/prometheus-k8s-operator/blob/6f2676c2674915777a694d8faa11df389a1b5670/.github/workflows/pull-request.yaml#L34).
Idk if pyright has the same issue.

**Review by rwcarlsen — APPROVED** [review]
I have the same comment/concern as @jnsgruk - maybe we should silence the warning and get rid of the casting.

File: pyproject.toml (hunk 2 above)
Comment (PietroPasotti) [reply pyproject.toml:39]: mmm. Let's add that just to be sure.

Comment (PietroPasotti) [reply pyproject.toml:39]: And that's how I discover that pyright supports python3.7+

Comment (PietroPasotti) [reply pyproject.toml:39]: Not sure if it's worth to support both mypy and pyright at this point, and use mypy for python 3.5. It looks like a lot more work.

**Pull request merged**